h/views/activity.py

```diff
@@ -296,17 +296,26 @@ def delete_lozenge(self):
     def toggle_tag_facet(self):
         return _toggle_tag_facet(self.request)
 
+    def _user_can_join(self):
+        """
+        Decide whether to show a page which will prompt the user to join.
+
+        :returns True if the user has permission to join or they are not logged in
+        and False otherwise.
+        """
+        if not self.request.user:
+            return True
+        return self.request.has_permission('join', self.group)
+
     def _check_access_permissions(self):
+        """
+        Check the read permission policy and if necessary redirect to the join page.
+        Otherwise if there is no polciy to read or join, raise a not found error.
+        """
```

Comment:
I'd try to avoid implementation details (check the read permission policy ... if there is no policy to read or join ... raise NotFound) in the docstring. The purpose of this function seems to be to:

1. Replace the page with a join page, if the user isn't authorized to read the group but does have permission to join it
2. Replace the page with a 404 page, if the user isn't authorized to either read or join the group
3. Do nothing if the user is already authorized to read the group

So if you want to write a docstring for this, I'd try to write one that describes just those _results_ of calling the function.

The fact that it's so awkward and difficult to write a concise docstring covering the purpose of this function of course shows that it's not a very cohesive function.

I'm not sure in what circumstances a group page ever 404s by the way. Does this ever even happen?